File: .claude/skills/startup-multi-device-analysis/scripts/factors_report.py
#!/usr/bin/env python3
"""Correlate external/on-device factors with SDK-init window slowness, outlier-first.

Consumes passN-factors.json files produced by outlier_factors.py + outlier_metrics.sql
(per-iteration: effective CPU MHz, cluster clocks/limits, thread-state with D-state
blocked_function, ART verify/class-load, lock contention, GC, in-process threads,
other-process CPU, binder counts, swap). Prints:

  1. pass-level factor means (what separates slow passes from fast passes)
  2. pooled per-iteration correlations vs window delta (within-pass, n=all)
  3. the outlier catalogue: every iteration >+4 ms vs pass median, with its factor vector
  4. extreme-outlier detail (top slowest overall): top blocked functions + top competitors
  5. D-state blocked_function totals (what main-thread IO waits actually are)

Usage: python3 factors_report.py <dir-with-passN-factors.json>
"""

import json
import os
import statistics
import sys

SLOW_DELTA_MS = 4.0
LITTLE_CPUS = {int(x) for x in os.environ.get("LITTLE_CPUS", "0,1,2,3").split(",")}


def main():
    d = sys.argv[1]
    passes = []
    for i in range(1, 20):
        p = os.path.join(d, f"pass{i}-factors.json")
        if not os.path.exists(p):
            break
        with open(p) as f:
            passes.append(json.load(f))
    if not passes:
        print("no passN-factors.json found", file=sys.stderr)
        return 1

    rows = []
    for pi, data in enumerate(passes):
        med = statistics.median([it["window_ms"] for it in data])
        for j, it in enumerate(data):
            rows.append(derive(it, pi + 1, j, med))

    n = len(rows)
    print(f"external-factor analysis: {len(passes)} passes, {n} iterations")
    print()

    # 1. pass-level means
    print("1. pass-level factor means (rates are CPU-ms per window-ms)")
    hdr = (f"  {'pass':>4}{'win p50':>9}{'run p50':>9}{'eff_mhz':>9}{'lim_cl1':>9}"
           f"{'sysserv rate':>13}{'sf rate':>9}{'other rate':>11}{'artload':>9}"
           f"{'d_io':>7}{'swap MB':>9}")
    print(hdr)
    for pi in range(1, len(passes) + 1):
        pr = [r for r in rows if r["pass"] == pi]
        print(f"  {pi:>4}"
              f"{statistics.median([r['window'] for r in pr]):>9.1f}"
              f"{statistics.median([r['run'] for r in pr]):>9.1f}"
              f"{mean0([r['eff_mhz'] for r in pr if r['eff_mhz']]):>9.0f}"
              f"{mean0([r['lim_cl1'] for r in pr if r['lim_cl1']]):>9.0f}"
              f"{statistics.mean([r['sysserver_rate'] for r in pr]):>13.3f}"
              f"{statistics.mean([r['sf_rate'] for r in pr]):>9.3f}"
              f"{statistics.mean([r['other_rate'] for r in pr]):>11.3f}"
              f"{statistics.mean([r['artload'] for r in pr]):>9.2f}"
              f"{statistics.mean([r['d_io'] for r in pr]):>7.2f}"
              f"{statistics.mean([r['swap'] for r in pr]) / 1e6:>9.1f}")
    print()

    # 2. pooled correlations vs delta
    print("2. pooled correlations with window delta (vs own pass median), n=%d" % n)
    deltas = [r["delta"] for r in rows]
    factors = [
        ("eff_mhz (weighted CPU clock)", "eff_mhz"),
        ("little share", "cl0_share"),
        ("system_server CPU rate", "sysserver_rate"),
        ("surfaceflinger CPU rate", "sf_rate"),
        ("other-proc CPU rate (excl idle)", "other_rate"),
        ("in-proc bg-thread CPU rate", "inproc_rate"),
        ("main-thread D/io ms", "d_io"),
        ("main-thread runnable ms", "rq"),
        ("ART class-load ms (in window)", "artload"),
        ("ART verify ms (in window)", "artverify"),
        ("lock contention ms", "lock"),
        ("GC slice ms (in proc)", "gc"),
        ("binder txns (main thread)", "binder"),
        ("process swap bytes", "swap"),
        ("MemAvailable", "memavail"),
    ]
    for label, key in factors:
        vals = [r[key] if r[key] is not None else 0 for r in rows]
        print(f"  {label:<36} r = {pearson(deltas, vals):>6.2f}")
    runs = [r["run"] for r in rows]
    print(f"  {'(sanity) run ms':<36} r = {pearson(deltas, runs):>6.2f}")
    print()

    # 3. outlier catalogue
    slows = sorted([r for r in rows if r["delta"] > SLOW_DELTA_MS],
                   key=lambda r: -r["delta"])
    print(f"3. outlier catalogue — {len(slows)} iterations with delta > +{SLOW_DELTA_MS:.0f} ms")
    print(f"  {'iter':>13}{'win':>7}{'Δ':>6}{'run':>7}{'MHz':>6}{'lit%':>6}"
          f"{'ss rate':>8}{'sf rate':>8}{'oth':>6}{'d_io':>6}{'rq':>5}"
          f"{'artload':>8}{'verify':>7}{'lock':>6}{'gc':>5}")
    for r in slows:
        print(f"  pass{r['pass']}:it{r['iter']:03d}"
              f"{r['window']:>7.1f}{r['delta']:>+6.1f}{r['run']:>7.1f}"
              f"{r['eff_mhz'] or 0:>6.0f}{r['cl0_share'] * 100:>6.0f}"
              f"{r['sysserver_rate']:>8.2f}{r['sf_rate']:>8.2f}{r['other_rate']:>6.2f}"
              f"{r['d_io']:>6.2f}{r['rq']:>5.1f}{r['artload']:>8.2f}"
              f"{r['artverify']:>7.2f}{r['lock']:>6.2f}{r['gc']:>5.2f}")
    med_fast = [r for r in rows if abs(r["delta"]) < 2]
    print(f"  --- baseline (|Δ|<2 ms, n={len(med_fast)}): "
          f"run {statistics.median([r['run'] for r in med_fast]):.1f}, "
          f"ss rate {statistics.mean([r['sysserver_rate'] for r in med_fast]):.2f}, "
          f"sf rate {statistics.mean([r['sf_rate'] for r in med_fast]):.2f}, "
          f"oth {statistics.mean([r['other_rate'] for r in med_fast]):.2f}, "
          f"d_io {statistics.median([r['d_io'] for r in med_fast]):.2f}, "
          f"artload {statistics.median([r['artload'] for r in med_fast]):.2f}, "
          f"verify {statistics.median([r['artverify'] for r in med_fast]):.2f}")
    print()

    # 4. extreme-outlier detail
    print("4. extreme outliers — top competitors and blocked functions")
    for r in sorted(rows, key=lambda r: -r["window"])[:8]:
        top_oth = sorted(r["othercpu"].items(), key=lambda kv: -kv[1])[:4]
        top_blk = sorted(r["blocked"].items(), key=lambda kv: -kv[1])[:3]
        oth = ", ".join(f"{k.split('/')[-1]} {v:.1f}" for k, v in top_oth)
        blk = ", ".join(f"{k or '?'} {v:.2f}" for k, v in top_blk) or "none"
        print(f"  pass{r['pass']}:it{r['iter']:03d} win {r['window']:.1f} "
              f"(Δ{r['delta']:+.1f})  competitors[ms]: {oth}")
        print(f"       blocked_on: {blk}")
    print()

    # 5. blocked-function totals
    print("5. main-thread D-state blocked_function totals across all iterations, ms")
    agg = {}
    for r in rows:
        for k, v in r["blocked"].items():
            agg[k or "?"] = agg.get(k or "?", 0) + v
    for k, v in sorted(agg.items(), key=lambda kv: -kv[1])[:15]:
        print(f"  {k:<44}{v:>9.1f}")
    return 0


def mean0(vals):
    if not vals:
        return 0.0
    return statistics.mean(vals)


def cl0_share(it):
    """Little-core share of window run time. outlier_metrics.sql partitions
    run_cl0_ms/run_cl1_ms at a fixed cpu<4 boundary (it does not know device topology);
    remap that fixed partition against LITTLE_CPUS (see device_probe.py) so the share
    still means "little cluster" on devices where the little cluster is cpu4-7."""
    r_cl0 = it.get("run_cl0_ms", 0)
    r_cl1 = it.get("run_cl1_ms", 0)
    tot = r_cl0 + r_cl1
    if not tot:
        return 0
    if all(c >= 4 for c in LITTLE_CPUS):
        return r_cl1 / tot
    return r_cl0 / tot


def derive(it, pnum, j, pass_med):
    st = it.get("states", {})
    run = st.get("Running", 0)
    d_io = sum(v for k, v in st.items()
               if k.startswith("D") or k.startswith("S+io"))
    rq = sum(v for k, v in st.items() if k.split(":")[0] in ("R", "R+"))
    blocked = {}
    for k, v in st.items():
        if k.startswith("D") and ":" in k:
            blocked[k.split(":", 1)[1]] = blocked.get(k.split(":", 1)[1], 0) + v
    win = it["window_ms"]
    oth = it.get("othercpu", {})
    other_total = sum(v for k, v in oth.items() if k != "swapper")
    ss = oth.get("system_server", 0)
    sf = oth.get("/system/bin/surfaceflinger", 0)
    inproc = sum(it.get("inproc", {}).values())
    return {
        "pass": pnum, "iter": j, "window": win, "delta": win - pass_med,
        "run": run, "d_io": d_io, "rq": rq, "blocked": blocked,
        "eff_mhz": it.get("eff_mhz"), "lim_cl1": it.get("freq_limit_cl1"),
        "cl0_share": cl0_share(it),
        "sysserver_rate": ss / win, "sf_rate": sf / win,
        "other_rate": other_total / win, "inproc_rate": inproc / win,
        "artload": it.get("art_classload_ms", 0),
        "artverify": it.get("art_verify_ms", 0),
        "lock": it.get("lock_contention_ms", 0),
        "gc": it.get("gc_slice_ms", 0),
        "binder": it.get("binder_txn_cnt", 0),
        "swap": it.get("mem_swap", 0) or 0,
        "memavail": it.get("mem_available", 0) or 0,
        "othercpu": oth,
    }


def pearson(xs, ys):
    mx, my = statistics.mean(xs), statistics.mean(ys)
    num = sum((x - mx) * (y - my) for x, y in zip(xs, ys))
    den = (sum((x - mx) ** 2 for x in xs) * sum((y - my) ** 2 for y in ys)) ** 0.5
    if den == 0:
        return float("nan")
    return num / den


if __name__ == "__main__":
    sys.exit(main())

File: startup-tools/src/main/kotlin/io/embrace/startup/analysis/FactorsReport.kt
package io.embrace.startup.analysis

import io.embrace.startup.analysis.OutlierFactors.Record
import io.embrace.startup.core.stats.Descriptive
import io.embrace.startup.core.stats.Quantile
import io.embrace.startup.core.text.PyFormat

/**
 * `factors_report.py`: correlate external/on-device factors with SDK-init window slowness,
 * outlier-first - pass-level factor means, pooled per-iteration correlations against the window
 * delta, the outlier catalogue, extreme-outlier detail and D-state blocked-function totals.
 *
 * `outlier_metrics.sql` partitions run time at a fixed cpu<4 boundary; [cl0Share] remaps that
 * partition against the device's real little cluster, exactly as the Python did.
 */
object FactorsReport {

    const val SLOW_DELTA_MS: Double = 4.0

    /** The Python's per-iteration `derive()` row. */
    data class Row(
        val pass: Int,
        val iter: Int,
        val window: Double,
        val delta: Double,
        val run: Double,
        val dIo: Double,
        val rq: Double,
        val blocked: Map<String, Double>,
        val effMhz: Double?,
        val limCl1: Double?,
        val cl0Share: Double,
        val sysserverRate: Double,
        val sfRate: Double,
        val otherRate: Double,
        val inprocRate: Double,
        val artload: Double,
        val artverify: Double,
        val lock: Double,
        val gc: Double,
        val binder: Double,
        val swap: Double,
        val memavail: Double,
        val othercpu: Map<String, Double>,
    )

    fun derive(rec: Record, pass: Int, iter: Int, passMedian: Double, littleCpus: Set<Int>): Row {
        val st = rec.states
        val win = requireNotNull(rec.windowMs) { "${rec.trace} has no window_ms" }
        val oth = rec.othercpu
        return Row(
            pass = pass,
            iter = iter,
            window = win,
            delta = win - passMedian,
            run = st["Running"] ?: 0.0,
            dIo = st.filterKeys { it.startsWith("D") || it.startsWith("S+io") }.values.sum(),
            rq = st.filterKeys { it.substringBefore(":") == "R" || it.substringBefore(":") == "R+" }.values.sum(),
            blocked = blockedFunctions(st),
            effMhz = rec.effMhz,
            limCl1 = rec.freqLimitCl1,
            cl0Share = cl0Share(rec, littleCpus),
            sysserverRate = (oth["system_server"] ?: 0.0) / win,
            sfRate = (oth["/system/bin/surfaceflinger"] ?: 0.0) / win,
            otherRate = oth.filterKeys { it != "swapper" }.values.sum() / win,
            inprocRate = rec.inproc.values.sum() / win,
            artload = rec.artClassloadMs ?: 0.0,
            artverify = rec.artVerifyMs ?: 0.0,
            lock = rec.lockContentionMs ?: 0.0,
            gc = rec.gcSliceMs ?: 0.0,
            binder = rec.binderTxnCnt ?: 0.0,
            swap = rec.memSwap ?: 0.0,
            memavail = rec.memAvailable ?: 0.0,
            othercpu = oth,
        )
    }

    /** D-state time by `blocked_function`, summed over the io/non-io variants of each state key. */
    private fun blockedFunctions(states: Map<String, Double>): Map<String, Double> {
        val blocked = LinkedHashMap<String, Double>()
        states.forEach { (k, v) ->
            if (k.startsWith("D") && ":" in k) {
                val fn = k.substringAfter(":")
                blocked[fn] = (blocked[fn] ?: 0.0) + v
            }
        }
        return blocked
    }

    fun cl0Share(rec: Record, littleCpus: Set<Int>): Double {
        val rCl0 = rec.runCl0Ms ?: 0.0
        val rCl1 = rec.runCl1Ms ?: 0.0
        val tot = rCl0 + rCl1
        if (tot == 0.0) return 0.0
        return if (littleCpus.all { cpu -> cpu >= FIXED_SPLIT }) rCl1 / tot else rCl0 / tot
    }

    fun rows(passes: List<List<Record>>, littleCpus: Set<Int>): List<Row> =
        passes.flatMapIndexed { pi, data ->
            val med = Quantile.median(data.map { requireNotNull(it.windowMs) }.sorted())
            data.mapIndexed { j, rec -> derive(rec, pi + 1, j, med, littleCpus) }
        }

    fun report(passes: List<List<Record>>, littleCpus: Set<Int> = VarianceAnalysis.DEFAULT_LITTLE_CPUS): String {
        val rows = rows(passes, littleCpus)
        val out = StringBuilder()
        out.line("external-factor analysis: ${passes.size} passes, ${rows.size} iterations")
        out.line()
        out.passMeans(rows, passes.size)
        out.correlations(rows)
        out.catalogue(rows)
        out.extremes(rows)
        out.blockedTotals(rows)
        return out.toString()
    }

    private fun StringBuilder.line(s: String = ""): StringBuilder = append(s).append('\n')

    private fun StringBuilder.passMeans(rows: List<Row>, passCount: Int) {
        line("1. pass-level factor means (rates are CPU-ms per window-ms)")
        line(
            "  ${"pass".padStart(W4)}${"win p50".padStart(W9)}${"run p50".padStart(W9)}${"eff_mhz".padStart(W9)}" +
                "${"lim_cl1".padStart(W9)}${"sysserv rate".padStart(W13)}${"sf rate".padStart(W9)}" +
                "${"other rate".padStart(W11)}${"artload".padStart(W9)}${"d_io".padStart(W7)}${"swap MB".padStart(W9)}",
        )
        for (pi in 1..passCount) {
            val pr = rows.filter { it.pass == pi }
            line(
                "  ${pi.toString().padStart(W4)}" +
                    f(median(pr.map { it.window }), 1, W9) +
                    f(median(pr.map { it.run }), 1, W9) +
                    f(mean0(pr.mapNotNull { it.effMhz }.filter { it != 0.0 }), 0, W9) +
                    f(mean0(pr.mapNotNull { it.limCl1 }.filter { it != 0.0 }), 0, W9) +
                    f(PyFormat.exactMean(pr.map { it.sysserverRate }), 3, W13) +
                    f(PyFormat.exactMean(pr.map { it.sfRate }), 3, W9) +
                    f(PyFormat.exactMean(pr.map { it.otherRate }), 3, W11) +
                    f(PyFormat.exactMean(pr.map { it.artload }), 2, W9) +
                    f(PyFormat.exactMean(pr.map { it.dIo }), 2, W7) +
                    f(PyFormat.exactMean(pr.map { it.swap }) / MEGA, 1, W9),
            )
        }
        line()
    }

    private fun StringBuilder.correlations(rows: List<Row>) {
        line("2. pooled correlations with window delta (vs own pass median), n=${rows.size}")
        val deltas = rows.map { it.delta }
        val factors: List<Pair<String, (Row) -> Double>> = listOf(
            "eff_mhz (weighted CPU clock)" to { r -> r.effMhz ?: 0.0 },
            "little share" to { r -> r.cl0Share },
            "system_server CPU rate" to { r -> r.sysserverRate },
            "surfaceflinger CPU rate" to { r -> r.sfRate },
            "other-proc CPU rate (excl idle)" to { r -> r.otherRate },
            "in-proc bg-thread CPU rate" to { r -> r.inprocRate },
            "main-thread D/io ms" to { r -> r.dIo },
            "main-thread runnable ms" to { r -> r.rq },
            "ART class-load ms (in window)" to { r -> r.artload },
            "ART verify ms (in window)" to { r -> r.artverify },
            "lock contention ms" to { r -> r.lock },
            "GC slice ms (in proc)" to { r -> r.gc },
            "binder txns (main thread)" to { r -> r.binder },
            "process swap bytes" to { r -> r.swap },
            "MemAvailable" to { r -> r.memavail },
        )
        factors.forEach { (label, get) ->
            line("  ${label.padEnd(W36)} r = ${f(Descriptive.pearson(deltas, rows.map(get)), 2, W6)}")
        }
        line("  ${"(sanity) run ms".padEnd(W36)} r = ${f(Descriptive.pearson(deltas, rows.map { it.run }), 2, W6)}")
        line()
    }

    private fun StringBuilder.catalogue(rows: List<Row>) {
        val slows = rows.filter { it.delta > SLOW_DELTA_MS }.sortedByDescending { it.delta }
        line("3. outlier catalogue — ${slows.size} iterations with delta > +${PyFormat.fixed(SLOW_DELTA_MS, 0)} ms")
        line(
            "  ${"iter".padStart(W13)}${"win".padStart(W7)}${"Δ".padStart(W6)}${"run".padStart(W7)}${"MHz".padStart(W6)}" +
                "${"lit%".padStart(W6)}${"ss rate".padStart(W8)}${"sf rate".padStart(W8)}${"oth".padStart(W6)}" +
                "${"d_io".padStart(W6)}${"rq".padStart(W5)}${"artload".padStart(W8)}${"verify".padStart(W7)}" +
                "${"lock".padStart(W6)}${"gc".padStart(W5)}",
        )
        slows.forEach { r ->
            line(
                "  pass${r.pass}:it${iter3(r.iter)}${f(r.window, 1, W7)}${PyFormat.fixedSigned(r.delta, 1).padStart(W6)}" +
                    "${f(r.run, 1, W7)}${f(r.effMhz ?: 0.0, 0, W6)}${f(r.cl0Share * PERCENT, 0, W6)}" +
                    "${f(r.sysserverRate, 2, W8)}${f(r.sfRate, 2, W8)}${f(r.otherRate, 2, W6)}${f(r.dIo, 2, W6)}" +
                    "${f(r.rq, 1, W5)}${f(r.artload, 2, W8)}${f(r.artverify, 2, W7)}${f(r.lock, 2, W6)}${f(r.gc, 2, W5)}",
            )
        }
        val medFast = rows.filter { kotlin.math.abs(it.delta) < BASELINE_BAND }
        line(
            "  --- baseline (|Δ|<2 ms, n=${medFast.size}): " +
                "run ${PyFormat.fixed(median(medFast.map { it.run }), 1)}, " +
                "ss rate ${PyFormat.fixed(meanOrNan(medFast.map { it.sysserverRate }), 2)}, " +
                "sf rate ${PyFormat.fixed(meanOrNan(medFast.map { it.sfRate }), 2)}, " +
                "oth ${PyFormat.fixed(meanOrNan(medFast.map { it.otherRate }), 2)}, " +
                "d_io ${PyFormat.fixed(median(medFast.map { it.dIo }), 2)}, " +
                "artload ${PyFormat.fixed(median(medFast.map { it.artload }), 2)}, " +
                "verify ${PyFormat.fixed(median(medFast.map { it.artverify }), 2)}",
        )
        line()
    }

    private fun StringBuilder.extremes(rows: List<Row>) {
        line("4. extreme outliers — top competitors and blocked functions")
        rows.sortedByDescending { it.window }.take(TOP_EXTREMES).forEach { r ->
            val topOth = r.othercpu.entries.sortedByDescending { it.value }.take(TOP_COMPETITORS)
            val topBlk = r.blocked.entries.sortedByDescending { it.value }.take(TOP_BLOCKED)
            val oth = topOth.joinToString(", ") { "${it.key.substringAfterLast('/')} ${PyFormat.fixed(it.value, 1)}" }
            val blk = topBlk.joinToString(", ") { "${it.key.ifEmpty { "?" }} ${PyFormat.fixed(it.value, 2)}" }.ifEmpty { "none" }
            line(
                "  pass${r.pass}:it${iter3(r.iter)} win ${PyFormat.fixed(r.window, 1)} " +
                    "(Δ${PyFormat.fixedSigned(r.delta, 1)})  competitors[ms]: $oth",
            )
            line("       blocked_on: $blk")
        }
        line()
    }

    private fun StringBuilder.blockedTotals(rows: List<Row>) {
        line("5. main-thread D-state blocked_function totals across all iterations, ms")
        val agg = LinkedHashMap<String, Double>()
        rows.forEach { r -> r.blocked.forEach { (k, v) -> agg[k.ifEmpty { "?" }] = (agg[k.ifEmpty { "?" }] ?: 0.0) + v } }
        agg.entries.sortedByDescending { it.value }.take(TOP_BLOCKED_TOTALS).forEach { (k, v) ->
            line("  ${k.padEnd(W44)}${f(v, 1, W9)}")
        }
    }

    /** `statistics.median`, or NaN on an empty list where the Python would have raised. */
    private fun median(values: List<Double>): Double =
        if (values.isEmpty()) Double.NaN else Quantile.median(values.sorted())

    private fun meanOrNan(values: List<Double>): Double =
        if (values.isEmpty()) Double.NaN else PyFormat.exactMean(values)

    private fun mean0(values: List<Double>): Double = if (values.isEmpty()) 0.0 else PyFormat.exactMean(values)

    private fun iter3(i: Int) = i.toString().padStart(3, '0')

    private fun f(x: Double, decimals: Int, width: Int) = PyFormat.fixed(x, decimals).padStart(width)

    private const val FIXED_SPLIT = 4
    private const val MEGA = 1e6
    private const val PERCENT = 100.0
    private const val BASELINE_BAND = 2.0
    private const val TOP_EXTREMES = 8
    private const val TOP_COMPETITORS = 4
    private const val TOP_BLOCKED = 3
    private const val TOP_BLOCKED_TOTALS = 15
    private const val W4 = 4
    private const val W5 = 5
    private const val W6 = 6
    private const val W7 = 7
    private const val W8 = 8
    private const val W9 = 9
    private const val W11 = 11
    private const val W13 = 13
    private const val W36 = 36
    private const val W44 = 44
}
